### PR TITLE
Openapi fix spec

### PR DESCRIPTION
**Pull request by TobiPeterG**
This switches to openapi-codegenerator and tries to patch the spec file with the patcher from netbox. Doesn't solve all issues yet.

### CODE REVIEW

**Comment by TobiPeterG** [comment]
The code now successfully compiles. Please check if the changes to the spec file are appropriate.

File: development/scripts/fix-spec.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+# Traverse schemas
+if 'components' in data and 'schemas' in data['components']:
+    for name, schema in data['components']['schemas'].items():
+        # Handle failover_strategy
+        if 'failover_strategy' in schema.get('properties', {}):
+            prop = schema['properties']['failover_strategy']
+            if 'default' in prop and prop['default'] is None:
+                print(f"Removing 'default: null' in {name}.failover_strategy")
+                del prop['default']
```
Comment (glennmatthews) [inline development/scripts/fix-spec.py:19]: For reference:

```yaml
        failover_strategy:
          type: object
          properties:
            value:
              type: string
              enum:
              - ''
              - active-active
              - active-passive
            label:
              type: string
              enum:
              - (unspecified)
              - Active/Active
              - Active/Passive
          default: null
```

Looks like the default should probably be the empty string, not null.

File: development/scripts/fix-spec.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                type_property = schema['properties']['type']
+                if 'properties' in type_property and 'value' in type_property['properties']:
+                    value_property = type_property['properties']['value']
+                    if 'enum' in value_property and set(value_property['enum']) == {'primary', 'redundant'}:
+                        print(f"Replacing complex 'type' field in PowerFeed")
+                        schema['properties']['type'] = {
+                            'type': 'string',
+                            'enum': ['primary', 'redundant'],
+                            'default': 'primary'
+                        }
```
Comment (glennmatthews) [inline development/scripts/fix-spec.py:34]: For reference:

```yaml
        type:
          type: object
          properties:
            value:
              type: string
              enum:
              - primary
              - redundant
            label:
              type: string
              enum:
              - Primary
              - Redundant
          default:
            value: primary
            label: Primary
```

File: development/scripts/fix-spec.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                supply_property = schema['properties']['supply']
+                if 'properties' in supply_property and 'value' in supply_property['properties']:
+                    value_property = supply_property['properties']['value']
+                    if 'enum' in value_property and set(value_property['enum']) == {'ac', 'dc'}:
+                        print(f"Replacing complex 'supply' field in PowerFeed")
+                        schema['properties']['supply'] = {
+                            'type': 'string',
+                            'enum': ['ac', 'dc'],
+                            'default': 'ac'
+                        }
```
Comment (glennmatthews) [inline development/scripts/fix-spec.py:47]: For reference:

```yaml
        supply:
          type: object
          properties:
            value:
              type: string
              enum:
              - ac
              - dc
            label:
              type: string
              enum:
              - AC
              - DC
          default:
            value: ac
            label: AC
```

File: development/scripts/fix-spec.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                phase_property = schema['properties']['phase']
+                if 'properties' in phase_property and 'value' in phase_property['properties']:
+                    value_property = phase_property['properties']['value']
+                    if 'enum' in value_property and set(value_property['enum']) == {'single-phase', 'three-phase'}:
+                        print(f"Replacing complex 'phase' field in PowerFeed")
+                        schema['properties']['phase'] = {
+                            'type': 'string',
+                            'enum': ['single-phase', 'three-phase'],
+                            'default': 'single-phase'
+                        }
```
Comment (glennmatthews) [inline development/scripts/fix-spec.py:60]: For reference:

```yaml
        phase:
          type: object
          properties:
            value:
              type: string
              enum:
              - single-phase
              - three-phase
            label:
              type: string
              enum:
              - Single phase
              - Three-phase
          default:
            value: single-phase
            label: Single phase
```

File: development/scripts/fix-spec.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            # Replace complex `type` object with a simpler string enum in Prefix
+            if 'type' in schema['properties']:
+                type_property = schema['properties']['type']
+                if 'properties' in type_property and 'value' in type_property['properties']:
+                    print(f"Replacing complex 'type' field in Prefix")
+                    schema['properties']['type'] = {
+                        'type': 'string',
+                        'enum': ['container', 'network', 'pool'],
+                        'default': 'network'
+                    }
```
Comment (glennmatthews) [inline development/scripts/fix-spec.py:73]: For reference:

```yaml
        type:
          type: object
          properties:
            value:
              type: string
              enum:
              - container
              - network
              - pool
            label:
              type: string
              enum:
              - Container
              - Network
              - Pool
          default:
            value: network
            label: Network
```

File: development/scripts/fix-spec.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                    
+        if 'properties' in schema:
+            # Remove "null" item from nullable enums
```
Comment (glennmatthews) [inline development/scripts/fix-spec.py:76]: Can you point to an example of this case? I'm not finding one in the schema at a glance.

File: development/scripts/fix-spec.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+            # Fix nullable types
+            nullable_types = [
+                'parent_device',
+                'primary_ip',
+            ]
+
+            for ntype in nullable_types:
+                if ntype in schema['properties']:
+                    schema['properties'][ntype]['nullable'] = True
```
Comment (glennmatthews) [inline development/scripts/fix-spec.py:91]: Looking at the schema, I'm seeing where `parent_device` already has `nullable: true` in the schema, and I'm not seeing any `primary_ip` properties (only `primary_ip4` and `primary_ip6`)

File: development/scripts/fix-spec.py (hunk 2 above)
Comment (glennmatthews) [reply development/scripts/fix-spec.py:34]: This is an intended asymmetry in the API - on responses from the server, `type` should be a dict with keys `value` and `label`, but on requests to the server, the client should just specify a string corresponding to the `value`. The `PowerFeed` type should only be used on the response side, on the request side see `WritablePowerFeedRequest`, `PatchedWritablePowerFeedRequest`, `BulkWritablePowerFeedRequest`, `PatchedBulkWritablePowerFeedRequest`

Same comment on the other enum fields below.

File: development/scripts/fix-spec.py (hunk 1 above)
Comment (glennmatthews) [reply development/scripts/fix-spec.py:19]: Per slack it's the latter case that's an issue. Probably both are wrong though.

Comment (glennmatthews) [reply development/scripts/fix-spec.py:19]: Possibly relevant: https://github.com/OpenAPITools/openapi-generator/issues/16775

Comment (glennmatthews) [reply development/scripts/fix-spec.py:19]: Not clear yet where the `null` is coming from here as both the model and the serializer define `blank=True` and `allow_blank=True` but not `null=True` or `allow_null=True` here.

Comment (TobiPeterG) [reply development/scripts/fix-spec.py:19]: So this is an issue of the specification, right?
Should I create an issue on nautobot's project?

File: development/scripts/fix-spec.py (hunk 2 above)
Comment (TobiPeterG) [reply development/scripts/fix-spec.py:34]: Is this something we need to change here then and/or something that needs to be changed in the specification?
How should go-nautobot handle this?

File: development/scripts/fix-spec.py (hunk 6 above)
Comment (TobiPeterG) [reply development/scripts/fix-spec.py:76]: This was kept from the netbox script.
I know that front_image and rear_image do not like the nullable property and some other properties as well.
I'll have a look at this tomorrow and check if this is really required.

File: development/scripts/fix-spec.py (hunk 7 above)
Comment (TobiPeterG) [reply development/scripts/fix-spec.py:91]: This was also kept from the netbox script, will check tomorrow if neccessary like above.

File: development/scripts/fix-spec.py (hunk 2 above)
Comment (glennmatthews) [reply development/scripts/fix-spec.py:34]: Not sure yet. I currently believe that the generated schema is correct but there may be some nuance I'm missing.

File: development/scripts/fix-spec.py (hunk 6 above)
Comment (TobiPeterG) [reply development/scripts/fix-spec.py]: This isn't required, so I removed this section

File: development/scripts/fix-spec.py (hunk 1 above)
Comment (TobiPeterG) [reply development/scripts/fix-spec.py:19]: Created an issue :)
https://github.com/nautobot/nautobot/issues/6157

File: .travis.yml
Comment (chadell) [inline .travis.yml:1]: whhy do we need this?

Comment (TobiPeterG) [reply .travis.yml:1]: It was auto-generated by the code generator

Comment (chadell) [reply .travis.yml:1]: I would remove it, as we don't use it

Comment (TobiPeterG) [reply .travis.yml:1]: But does it hurt?

Comment (chadell) [inline .travis.yml:1]: could we remove this?

Comment (TobiPeterG) [reply .travis.yml:1]: I will remove it

Comment (TobiPeterG) [reply .travis.yml:1]: Removed

**Comment by TobiPeterG** [comment]
@chadell Is there something else that should be changed? :)

File: README.md [hunk 8]
```diff
@@ -1,131 +1,2469 @@
-# go-nautobot
+# Go API client for nautobot
 
-## :warning: Disclaimer :warning:
```
Comment (chadell) [inline README.md:3]: I want to keep the disclaimer until the project is more mature

File: README.md [hunk 9]
```diff
@@ -1,131 +1,2469 @@
...
+For using other server than the one defined on index 0 set context value `nautobot.ContextServerIndex` of type `int`.
 
-   ```bash
-   $ oapi-codegen --config oapi-config.yml swagger.yaml
-   ```
+```go
+ctx := context.WithValue(context.Background(), nautobot.ContextServerIndex, 1)
+```
 
-## Release Versioning
```
Comment (chadell) [inline README.md:39]: I would like to keep a documentation of the release logic

File: README.md [hunk 10]
```diff
@@ -1,131 +1,2469 @@
...
+})
+ctx = context.WithValue(context.Background(), nautobot.ContextOperationServerVariables, map[string]map[string]string{
+	"{classname}Service.{nickname}": {
+		"port": "8443",
+	},
+})
+```
 
-## How to use `go-nautobot` package
+## Documentation for API Endpoints
```
Comment (chadell) [inline README.md:75]: move this part to the end

File: README.md [hunk 11]
```diff
@@ -1,131 +1,2469 @@
...
+*VirtualizationAPI* | [**VirtualizationVirtualMachinesList**](docs/VirtualizationAPI.md#virtualizationvirtualmachineslist) | **Get** /virtualization/virtual-machines/ | 
+*VirtualizationAPI* | [**VirtualizationVirtualMachinesNotesCreate**](docs/VirtualizationAPI.md#virtualizationvirtualmachinesnotescreate) | **Post** /virtualization/virtual-machines/{id}/notes/ | 
+*VirtualizationAPI* | [**VirtualizationVirtualMachinesNotesList**](docs/VirtualizationAPI.md#virtualizationvirtualmachinesnoteslist) | **Get** /virtualization/virtual-machines/{id}/notes/ | 
+*VirtualizationAPI* | [**VirtualizationVirtualMachinesPartialUpdate**](docs/VirtualizationAPI.md#virtualizationvirtualmachinespartialupdate) | **Patch** /virtualization/virtual-machines/{id}/ | 
+*VirtualizationAPI* | [**VirtualizationVirtualMachinesRetrieve**](docs/VirtualizationAPI.md#virtualizationvirtualmachinesretrieve) | **Get** /virtualization/virtual-machines/{id}/ | 
+*VirtualizationAPI* | [**VirtualizationVirtualMachinesUpdate**](docs/VirtualizationAPI.md#virtualizationvirtualmachinesupdate) | **Put** /virtualization/virtual-machines/{id}/ | 
 
-```bash
-$ go get github.com/nautobot/go-nautobot/pkg/nautobot@v1.5.8-beta
```
Comment (chadell) [inline README.md:65]: I think it is useful

File: README.md [hunk 12]
```diff
@@ -1,131 +1,2469 @@
...
+		context.Background(),
+		nautobot.ContextAPIKeys,
+		map[string]nautobot.APIKey{
+			"sessionid": {Key: "API_KEY_STRING"},
+		},
+	)
+r, err := client.Service.Operation(auth, args)
+```
 
-- [Network Automation with Go Book](https://github.com/PacktPublishing/Network-Automation-with-Go/blob/main/ch06/nautobot/main.go)
```
Comment (chadell) [inline README.md:111]: I think that having an example is great to get started, and adding the reference to this book it's good too even though is using an old version

File: README.md [hunk 13]
```diff
@@ -1,131 +1,2469 @@
...
+		map[string]nautobot.APIKey{
+			"Authorization": {Key: "API_KEY_STRING"},
+		},
+	)
+r, err := client.Service.Operation(auth, args)
 ```
 
-> Hint: If you get a build fail during testing, check that you are not limiting container memory to 2GB. Upgrade to 4GB.
 
-### Trigger the release manually
```
Comment (chadell) [inline README.md:125]: please, keep the memory warning if you als0 need it

and also keep the small section about how to run the workflow in GitHub actions to trigger the release

File: README.md [hunk 14]
```diff
@@ -1,131 +1,2469 @@
...
+* `PtrInt`
+* `PtrInt32`
+* `PtrInt64`
+* `PtrFloat`
+* `PtrFloat32`
+* `PtrFloat64`
+* `PtrString`
+* `PtrTime`
+
+## Author
```
Comment (chadell) [inline README.md:2466]: please, add yourself as the big contributor!

File: README.md (hunk 8 above)
Comment (TobiPeterG) [reply README.md:3]: Done :)

File: README.md (hunk 9 above)
Comment (TobiPeterG) [reply README.md:39]: Done

File: README.md (hunk 10 above)
Comment (TobiPeterG) [reply README.md:75]: I added a part for documentation

File: README.md (hunk 11 above)
Comment (TobiPeterG) [reply README.md:65]: Added

File: README.md (hunk 12 above)
Comment (TobiPeterG) [reply README.md:111]: I changed the reference to the terraform provider, as that will use the new go-nautobot code :)

File: README.md (hunk 13 above)
Comment (TobiPeterG) [reply README.md:125]: Done

File: README.md (hunk 14 above)
Comment (TobiPeterG) [reply README.md:2466]: Done, added an Author section :)

File: .README.md [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+
+func check(err error) {
+	if err != nil {
+		panic(err)
+	}
+}
+
+func main() {
+	token := "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+	check(err)
```
Comment (chadell) [inline .README.md:93]: `err` is not defined at this point, please remove

File: .README.md [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+	
+	config := nb.NewConfiguration()
+	config.Servers[0].URL = "localhost"
```
Comment (chadell) [inline .README.md:96]: `demo.nautobot.com` ?

**Review by chadell — APPROVED** [review]
LGTM

**Pull request merged**